From 3f5fca5eb96a91463bade6f709485f0fc4427da4 Mon Sep 17 00:00:00 2001
From: Jon Turney <jon.turney@dronecode.org.uk>
Date: Fri, 16 Oct 2026 11:04:37 +0530
Subject: [PATCH 10/11] ld: Don't emit a reference to _pei386_runtime_relocator
 for PE+

ld synthesises a reference to _pei386_runtime_relocator to satisfy a
non-functional stub in the Cygwin runtime.  The MinGW runtime still has real
code there and calls it explicitly at startup; the Cygwin stub exists only to
satisfy this linker-generated reference.

Stop emitting the reference on PE+ targets, so the stub can be dropped on new
architectures such as aarch64 where there is nothing to stay compatible with.
i386 keeps the existing behaviour out of an abundance of caution.

Taken from the patch posted to binutils@ on 2026-07-02:
https://sourceware.org/pipermail/binutils/2026-July/149962.html
Not yet merged upstream; drop this once it is.

---
 ld/pe-dll.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/ld/pe-dll.c b/ld/pe-dll.c
index 4ff3ea0ff20..3bde5f6131f 100644
--- a/ld/pe-dll.c
+++ b/ld/pe-dll.c
@@ -170,7 +170,9 @@ static struct bfd_section *edata_s, *reloc_s;
 static unsigned char *edata_d, *reloc_d;
 static unsigned char *reloc_d = NULL;
 static size_t edata_sz, reloc_sz = 0;
+#ifndef pe_use_plus
 static int runtime_pseudo_relocs_created = 0;
+#endif
 static bool runtime_pseudp_reloc_v2_init = false;
 
 typedef struct
@@ -2798,6 +2800,7 @@ make_runtime_pseudo_reloc (const char *name ATTRIBUTE_UNUSED,
   return abfd;
 }
 
+#ifndef pe_use_plus
 /*	.section	.rdata
 	.rva		__pei386_runtime_relocator  */
 
@@ -2841,6 +2844,7 @@ pe_create_runtime_relocator_reference (bfd *parent)
   bfd_make_readable (abfd);
   return abfd;
 }
+#endif
 
 static void
 aarch64_make_jump_stub(const char* symbol_name, bfd* parent)
@@ -3067,11 +3071,15 @@ pe_create_import_fixup (arelent *rel, asection *s, bfd_vma addend, char *name,
 				     link_info.output_bfd);
       add_bfd_to_link (b, bfd_get_filename (b), &link_info);
 
+#ifndef pe_use_plus
+      /* We think this is entirely useless, but retain it for i386 out of
+         an abundance of caution.  */
       if (runtime_pseudo_relocs_created++ == 0)
 	{
 	  b = pe_create_runtime_relocator_reference (link_info.output_bfd);
 	  add_bfd_to_link (b, bfd_get_filename (b), &link_info);
 	}
+#endif
     }
 
   else if (addend != 0)
-- 
2.51.0.windows.1

